### PR TITLE
Integrate timeloop

### PR DESCRIPTION
**Pull request by Jared-Chan**
# Changes and new stuff
- .devcontainer/Dockerfile
  - add dependencies of Timeloop
  - add stages for installing and configuring Timeloop and Accelergy
- .gitignore
  - ignore timeloop output
- README.md
  - add Timeloop info
- src/main.py
  - add option to run timeloop-mapper
  - add steps to parse output after running timeloop-mapper
- src/timeloop_tools.py
  - functions to run timeloop-mapper and parse timeloop output
  - they are called by main.py 
- timeloop/
  - Timeloop input

## Test
`python src/main.py ifrneten_manual -m`

Expected output (the last two lines):
```
Stats: total_area=1,942,452.4353820002um^2, gflops=565.06, utilization=0.82
cycles=1,166,963, latency=1.166963ms
```

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -159,6 +170,67 @@ def main():
         latency = (macs / EYERISSV2_MACS) / EYERISSV2_FREQ
         print(f'Stats: {macs=:,}, {params=:,}, {latency=:,}ms')
 
+    if args.map:
```
Comment (patrick-5546) [inline src/main.py:168]: Can you put this in its own file? Besides the if-else to select the model, would like each other if statement to only be a couple lines long

File: tl_setup.sh [hunk 2]
```diff
@@ -0,0 +1,166 @@
+# Build Timeloop Infrastructure
+printf "\n# Set up Timeloop Infrastructure\n"
+
+pushd $FRUC_WORKSPACE
+
+BIN_DIR=/usr/local/bin
+BUILD_DIR=/usr/local/src
+LIB_DIR=/usr/local/lib
+SHARE_DIR=/usr/local/share
+INCLUDE_DIR=/usr/local/include
```
Comment (patrick-5546) [inline tl_setup.sh:10]: The only thing about doing this is that when you have to rebuild from scratch every time. 2 solution I see:
1. Build dir in /workspaces/fruc_workspace: this is like I did from python, installing everything in a `.venv/`
2. Run in dockerfile

File: tl_setup.sh [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+#   3. In the accelergy config, add paths to plug ins and component libs
+#   4. Expect a permission error when running timeloop-mapper regarding 
+#      creating the 'out' dir
+
+# Expected warning: 
+# WARNING: Running pip as the 'root' user can result in broken permissions and con
+# flicting behaviour with the system package manager. It is recommended to use a v
+# irtual environment instead: https://pip.pypa.io/warnings/venv
+
+python3 -m pip install setuptools
```
Comment (patrick-5546) [inline tl_setup.sh:106]: Getting running pip as the root user warnings. If this is staying in this script (as opposed to moving to dockerfile), should do something similar to `setup.sh` which sources the virtual environment and installs stuff there

**Review by patrick-5546 — COMMENTED** [review]
Fantastic work jared, thanks so much! Left a couple comments below

File: src/main.py (hunk 1 above)
Comment (Jared-Chan) [reply src/main.py:170]: Yep, it is now in timeloop_tool.py

File: tl_setup.sh (hunk 2 above)
Comment (Jared-Chan) [reply tl_setup.sh]: Turned out doing this in the dockerfile isn't too difficult, the whole tl_setup.sh is moved to the dockerfile now.

File: src/timeloop_tools.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        bash_command += "arch_dense/*.yaml arch_dense/components/* "
+    bash_command += "constraint/* mapper/* "
+    bash_command += f"{problem_f} "
+    bash_command += "-o out"
+    # Run timeloop-mapper
+    print(bash_command)
+    os.system(bash_command)
+
+
+def parse_mapper_output() -> float | int | float | float:
```
Comment (patrick-5546) [inline src/timeloop_tools.py:29]: If you pass in out dir as argument, no need to have duplicate constant in this file `FRUC_WORKSPACE`

File: src/timeloop_tools.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                l_utilization = f.readline()
+                l_cycles = f.readline()
+                cycles += int(l_cycles[l_cycles.find('Cycles: ') + 8:])
+                gflops += float(l_gflops[l_gflops.find('GFLOPs (@1GHz): ') + 16:])
+                util_str = l_utilization[l_utilization.find('Utilization: ') + 13:]
+                utilization += float(util_str)
+                break
+            stat_l = f.readline()
+    
+    return total_area, cycles, utilization, gflops
```
Comment (patrick-5546) [inline src/timeloop_tools.py:66]: Empty line at end of file thanks

**Pull request merged**